File: 2020/3.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "input_3.txt"
    lines = readInput(filename)
    pattern = Dict{Tuple{Int, Int}, Int}() 
    for y in eachindex(lines)
        for x in 1:length(lines[1])
            if lines[y][x] == '#'
                pattern[(x-1, y-1)] = 1
            end
        end
    end
    return length(lines[1]), length(lines), pattern
end

function traverse(dx::Int, dy::Int, xmax::Int, ymax::Int, pattern::Dict{Tuple{Int, Int}, Int})
    y = 0
    x = 0
    trees = 0
    while true
        y += dy
        if y >= ymax
            break
        end
        x += dx
        x %= xmax
        if haskey(pattern, (x, y))
            trees += 1
        end
    end
    return trees
end

function part1(xmax::Int, ymax::Int, pattern::Dict{Tuple{Int, Int}, Int})
    trees = traverse(3, 1, xmax, ymax, pattern)
    println("🎄 Part 1: $(trees)")
end

function part2(xmax::Int, ymax::Int, pattern::Dict{Tuple{Int, Int}, Int})
    slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    trees = Int[]
    for s in slopes
        push!(trees, traverse(s..., xmax, ymax, pattern))
    end
    println("🎄🎅 Part 2: $(prod(trees))")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 3         ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

xmax, ymax, inputs = loadInput()
@time part1(xmax, ymax, inputs)
@time part2(xmax, ymax, inputs)

File: 2020/3.py
import time
import numpy as np
from utils import readInput

def loadInput():
    lines = readInput("input_3.txt")
    return lines

def part1(lines):
    y = 0
    x = 0
    trees = 0
    N = len(lines)
    Nx = len(lines[0])
    while True:
        y += 1
        if y >= N:
            break
        x += 3
        x = x % Nx
        if lines[y][x] == "#":
            trees += 1
    print ("🎄 Part 1: {}".format(trees))

def part2(lines):
    slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    trees = []

    N = len(lines)
    Nx = len(lines[0])

    for slope in slopes:
        y = 0
        x = 0
        tree = 0
        while True:
            y += slope[1]
            if y >= N:
                break
            x += slope[0]
            x = x % Nx

            if lines[y][x] == "#":
                tree += 1
        trees.append(tree)
    
    print ("🎄🎅 Part 2: {}".format(np.prod(trees)))

print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
print("⛄        Day 3         ⛄")
print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()

t0 = time.time()
part1(inputs)
print ("Time: {:.5f}".format(time.time()-t0))

t0 = time.time()
part2(inputs)
print ("Time: {:.5f}".format(time.time()-t0))
